Add authentication/authorization docs for v11 (#3830)
ChilliCream/hotchocolate: website/src/docs/hotchocolate/security/authorization.md

---
title: Authorization
---

import { ExampleTabs } from "../../../components/mdx/example-tabs"

Authorization allows us to determine a user's permissions within our system. We can for example limit access to resources or only allow certain users to execute specific mutations.

Authentication is a prerequisite of Authorization, as we first need to validate a user's "authenticity" before we can evaluate his authorization claims.

[Learn how to setup authentication](/docs/hotchocolate/security/authentication)

# Setup

After we have successfully setup authentication, there are only a few things left to do.

1. Register the necessary ASP.NET Core services

```csharp
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddAuthorization();

        // Omitted code for brevity

        services
            .AddGraphQLServer()
            .AddAuthorization()
            .AddQueryType<Query>();
    }
}
```

> ⚠️ Note: We need to call `AddAuthorization()` on the `IServiceCollection`, to register the services needed by ASP.NET Core, and on the `IRequestExecutorBuilder` to register the `@authorize` directive and middleware.

> 2. Register the `UseAuthorization` middleware with the request pipeline

Reviewer: ```suggestion
2. Register the ASP.NET Core authorization middleware with the request pipeline by calling `UseAuthorization`
```

```csharp
public class Startup
{
    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        app.UseRouting();

        app.UseAuthentication();
        app.UseAuthorization();

        app.UseEndpoints(endpoints =>
        {
            endpoints.MapGraphQL();
        });
    }
}
```

# Usage

At the core of authorization with Hot Chocolate is the `@authorize` directive. It can be applied to fields and types to denote that they require authorization.

> If the `@authorize` directive is specified on a type, it is applied to all fields of that type. Specified on an individual field the directive precedences the one on the type.

Reviewer: ```suggestion
If the `@authorize` directive is specified on a type, it is applied to all fields of that type. Specified on an individual field the directive overrules the one on the type.
```

I think it's worth mentioning here in an aside/note that it applies the type-level authorization by essentially annotating every field, so the authorization logic will be run once for each field.

cc: @michaelstaib

Author: Do you like this better?
![image](https://user-images.githubusercontent.com/45513122/124011154-fb2e9300-d9df-11eb-8ba6-c8805f5b6b68.png)

    [Authorize]
    public Address Address { get; set; }
}
```

> ⚠️ Note: We need to use the `HotChocolate.AspNetCore.AuthorizationAttribute` instead of the `Microsoft.AspNetCore.AuthorizationAttribute`.

</ExampleTabs.Annotation>
<ExampleTabs.Code>

```csharp
public class UserType : ObjectType<User>
{
    protected override void Configure(IObjectTypeDescriptor<User> descriptor)
    {
        descriptor.Authorize();

        descriptor.Field(f => f.Address).Authorize();
    }
}
```

</ExampleTabs.Code>
<ExampleTabs.Schema>

```sdl
type User @authorize {
  name: String!
  address: Address! @authorize
}
```

</ExampleTabs.Schema>
</ExampleTabs>

If we do not specify any arguments to the `@authorize` directive, it will only enforce that the requestor is authenticated, nothing more. If he is not and he tries to access an authorized field, a GraphQL error will be raised and the field result set to `null`.

## Roles

Roles provide a very intuitive way of dividing our users into groups with different access rights.

When building our `ClaimsPrincipal`, we just have to add one or more role claims.

```csharp
claims.Add(new Claim(ClaimTypes.Role, "Administrator"));
```

> We can then check, whether an authenticated user has these role claims.

Reviewer: ```suggestion
We can then check whether an authenticated user has these role claims.
```

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp
[Authorize(Roles = new [] { "Guest", "Administrator" })]
public class User
{
    public string Name { get; set; }

    [Authorize(Roles = new[] { "Administrator" })]
    public Address Address { get; set; }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

```csharp
public class UserType : ObjectType<User>
{
    protected override Configure(IObjectTypeDescriptor<User> descriptor)
    {
        descriptor.Authorize(new[] { "Guest", "Administrator" });

        descriptor.Field(t => t.Address).Authorize(new[] { "Administrator" });
    }
}
```

</ExampleTabs.Code>
<ExampleTabs.Schema>

```sdl
type User @authorize(roles: [ "Guest", "Administrator" ]) {
  name: String!
  address: Address! @authorize(roles: "Administrator")
}
```

</ExampleTabs.Schema>
</ExampleTabs>

> ⚠️ Note: If multiple roles are specified, a user only has to match one of the specified roles, in order to be able to execute the resolver.

[Learn more about role-based authorization in ASP.NET Core](https://docs.microsoft.com/aspnet/core/security/authorization/roles)

> ## Policies

Reviewer: It's probably worth adding a similar link for policies to the asp.net core docs as done just above for roles.

Author: I had one initially, must've gotten lost, while I was shifting stuff around. Good catch!

        services.AddSingleton<IAuthorizationHandler, MinimumAgeHandler>();

        // Omitted code for brevity

        services
            .AddGraphQLServer()
            .AddAuthorization()
            .AddQueryType<Query>();
    }
}
```

We can then use these policies to restrict access to our fields.

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp
[Authorize(Policy = "AllEmployees")]
public class User
{
    public string Name { get; }

    [Authorize(Policy = "SalesDepartment")]
    public Address Address { get; }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

```csharp
public class UserType : ObjectType<User>
{
    protected override Configure(IObjectTypeDescriptor<User> descriptor)
    {
        descriptor.Authorize("AllEmployees");

        descriptor.Field(t => t.Address).Authorize("SalesDepartment");
    }
}
```

</ExampleTabs.Code>
<ExampleTabs.Schema>

```sdl
type User @authorize(policy: "AllEmployees") {
  name: String!
  address: Address! @authorize(policy: "SalesDepartment")
}
```

</ExampleTabs.Schema>
</ExampleTabs>

This essentially uses the provided policy and runs it against the `ClaimsPrinciple` that is associated with the current request.

> The `@authorize` directive is also repeatable, which means that we are able to chain the directive and a user is only allowed to access the field, if he meets all of the specified conditions.

Reviewer: ```suggestion
The `@authorize` directive is also repeatable, which means that we are able to chain the directive and a user is only allowed to access the field if they meet all of the specified conditions.
```

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp
[Authorize(Policy = "AtLeast21")]
[Authorize(Policy = "HasCountry")]
public class User
{
    public string Name { get; set; }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

```csharp
public class UserType : ObjectType<User>
{
    protected override Configure(IObjectTypeDescriptor<User> descriptor)
    {
        descriptor
            .Authorize("AtLeast21")
            .Authorize("HasCountry");
    }
}
```

</ExampleTabs.Code>
<ExampleTabs.Schema>

```sdl
type User
	@authorize(policy: "AtLeast21")
	@authorize(policy: "HasCountry") {
	name: String!
}
```

</ExampleTabs.Schema>
</ExampleTabs>

### IResolverContext within an AuthorizationHandler

If we need to, we can also access the `IResolverContext` in our `AuthorizationHandler`.

```csharp
public class MinimumAgeHandler
    : AuthorizationHandler<MinimumAgeRequirement, IResolverContext>
{
    protected override Task HandleRequirementAsync(
        AuthorizationHandlerContext context,
        MinimumAgeRequirement requirement,
        IResolverContext resolverContext)
    {
        // Omitted code for brevity
    }
}
```

## Global authorization

> We can also apply authorization to our entire GraphQL endpoint, by calling `RequireAuthorization()` on the `GraphQLEndpointConventionBuilder`.

Reviewer: ```suggestion
We can also apply authorization to our entire GraphQL endpoint. To do this,  simply call `RequireAuthorization()` on the `GraphQLEndpointConventionBuilder`.
```

```csharp
public class Startup
{
    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        app.UseRouting();

        app.UseAuthentication();
        app.UseAuthorization();

        app.UseEndpoints(endpoints =>
        {
            endpoints.MapGraphQL().RequireAuthorization();
        });
    }
}
```

> ⚠️ Note: This will also block unauthenticated access to GraphQL IDEs hosted on that endpoint, like Banana Cake Pop.

This method also accepts [roles](#roles) and [policies](#policies) as arguments, similiar to the `Authorize` attribute / methods.

<!-- ## Adding claims dynamically

> TODO: Rework and test

Reviewer: ping
